### PR TITLE
Fix flow typing for BigNumber.js

### PR DESCRIPTION
**Pull request by mdjastrzebski**
Scope:
* Added flow-typed definitions for BigNumber.js (before that it was treated as `any`)
* Fixed most type errors
* Ignored some remaining errors without clear resolution

Types of changes:
* type changes - these are safe, as they do not change the runtime behaviour
* code changes - this could cause runtime issues, so I've only changed code that seem safe to change  but **please check them carefully**.

### CODE REVIEW

File: src/actions/transactionEstimateActions.js [hunk 1]
```diff
@@ -129,7 +129,7 @@ export const estimateTransactionAction = (
       feeInfo = buildTxFeeInfo(estimated, useGasToken);
     }
 
-    if (!errorMessage && feeInfo && !feeInfo.fee.gt(0)) {
+    if (!errorMessage && feeInfo && !feeInfo.fee?.gt(0)) {
```
Comment (mdjastrzebski) [inline src/actions/transactionEstimateActions.js:132]: Code analysis seem to indicate that null `fee` means that estimation has failed.

File: src/screens/Exchange/utils.js [hunk 2]
```diff
@@ -200,7 +200,7 @@ export const getHeaderRightItems = (
 const isEnoughAssetBalance = (assetBalance: ?string, amount: string): boolean => {
   try {
     const amountBN = new BigNumber(amount);
-    const balanceBN = new BigNumber(assetBalance);
+    const balanceBN = new BigNumber(assetBalance ?? 0);
```
Comment (mdjastrzebski) [inline src/screens/Exchange/utils.js:203]: This change assumes that missing asset balance is equal to balance of 0 which will lead to returning false for all non-zero `amount` values. Previously this was BigNumber(NaN)

File: src/screens/SendCollectible/SendCollectibleConfirm.js [hunk 3]
```diff
@@ -164,7 +164,7 @@ const SendCollectibleConfirm = ({
 
   if (feeInfo) {
     // rinkeby testnet fee check
-    const txFee = utils.formatEther(feeInfo.fee.toString());
+    const txFee = utils.formatEther(feeInfo.fee?.toString() ?? 0);
```
Comment (mdjastrzebski) [inline src/screens/SendCollectible/SendCollectibleConfirm.js:167]: Previously this threw error iff `fee` == `null`

File: src/services/uniswap.js [hunk 4]
```diff
@@ -163,7 +163,7 @@ export const getUniswapOffer = async (
   const askRate = expectedOutput.dividedBy(quantityBN);
   const allowanceSet = await getAllowanceSet(clientAddress, fromAssetParsed);
 
-  return parseOffer(fromAssetParsed, toAssetParsed, allowanceSet, askRate, PROVIDER_UNISWAP);
+  return parseOffer(fromAssetParsed, toAssetParsed, allowanceSet, askRate.toFixed(), PROVIDER_UNISWAP);
```
Comment (mdjastrzebski) [inline src/services/uniswap.js:166]: All other parseOffer invocations passed string, produced by `toFixed()` or `toFixed(n)` calls.

File: src/utils/assets.js [hunk 5]
```diff
@@ -199,7 +199,7 @@ export const isEnoughBalanceForTransactionFee = (
     }
   }
 
-  const txFeeInWeiBN = new BigNumber(txFeeInWei.toString()); // compatibility
+  const txFeeInWeiBN = new BigNumber(txFeeInWei?.toString() ?? 0); // compatibility
```
Comment (mdjastrzebski) [inline src/utils/assets.js:202]: Otherwise would throw for null txFeeInWei

File: src/services/smartWallet.js [hunk 6]
```diff
@@ -299,7 +299,7 @@ class SmartWallet {
       });
   }
 
-  getAccountStakedAmount(tokenAddress: ?string): BigNumber {
+  getAccountStakedAmount(tokenAddress: ?string): Promise<BigNumber> | BigNumber {
```
Comment (jdnmr) [inline src/services/smartWallet.js:302]: Isn't `Promise<BigNumber>` enough?

File: src/utils/assets.js [hunk 7]
```diff
@@ -137,7 +137,7 @@ export const getFormattedRate = (
 export const calculateMaxAmount = (
   token: string,
   balance: number | string,
-  txFeeInWei: BigNumber,
+  txFeeInWei: ?(BigNumber | string | number),
```
Comment (jdnmr) [inline src/utils/assets.js:140]: We have a type `Value` for this in `utils/common`

File: src/utils/assets.js [hunk 8]
```diff
@@ -164,7 +164,7 @@ export const calculateMaxAmount = (
 export const isEnoughBalanceForTransactionFee = (
   balances: Balances,
   transaction: {
-    txFeeInWei: number,
+    txFeeInWei: ?(BigNumber | string | number),
```
Comment (jdnmr) [inline src/utils/assets.js:167]: You can use `Value` from `utils/common`

File: src/utils/common.js [hunk 9]
```diff
@@ -182,7 +182,7 @@ export const formatMoney = (
   return (c ? num.replace('.', c) : num).replace(new RegExp(re, 'g'), `$&${s || ','}`);
 };
 
-export const parseNumber = (amount: string = '0') => {
+export const parseNumber = (amount: BigNumber | string | number = '0') => {
```
Comment (jdnmr) [inline src/utils/common.js:185]: `Value`... and in other places in this file too :-)

**Review by jdnmr — APPROVED** [review]
LG, just a few small comments

File: src/services/smartWallet.js (hunk 6 above)
Comment (mdjastrzebski) [reply src/services/smartWallet.js:302]: No, as next line there is an early return ` if (!tokenAddress) return new BigNumber(0);` not wrapped in promise.

File: src/utils/assets.js (hunk 7 above)
Comment (mdjastrzebski) [reply src/utils/assets.js:140]: Fixed

**Pull request merged**